src/java/org/apache/cassandra/service/disk/usage/DiskUsageMonitor.java

```diff
@@ -0,0 +1,178 @@
+        this.failValueSupplier = () -> config.disk_usage_percentage_failure_threshold;
+        this.dataDirectoriesSupplier = () -> Directories.dataDirectories;
+    }
+
+    /**
+     * Start monitoring local disk usage and call notifier when local disk usage state changed.
+     */
+    public void start(Consumer<DiskUsageState> notifier)
+    {
+        // start the scheduler regardless guardrail is enabled, so we can enable it later without a restart
+        ScheduledExecutors.scheduledTasks.scheduleAtFixedRate(() -> {
+
+            if (!Guardrails.localDiskUsage.enabled())
+                return;
+
+            updateLocalState(getDiskUsage(), notifier);
+        }, 0, MONITOR_INTERVAL, TimeUnit.MILLISECONDS);
+    }
+
+    @VisibleForTesting
+    public void updateLocalState(double usageRatio, Consumer<DiskUsageState> notifier)
+    {
+        double percentage = usageRatio * 100;
+        long percentageCeiling = (long) Math.ceil(percentage);
+
+        DiskUsageState state = getState(percentageCeiling);
+
+        Guardrails.localDiskUsage.guard(percentageCeiling, state.toString(), false);
+
+        // if state remains unchanged, no need to notify peers
+        if (state == localState)
+            return;
+
+        localState = state;
+        notifier.accept(state);
+    }
+
+    /**
+     * @return local node disk usage state
+     */
+    @VisibleForTesting
+    public DiskUsageState state()
+    {
+        return localState;
+    }
+
+    /**
+     * @return disk usage (including all memtable sizes) ratio
+     */
+    private double getDiskUsage()
+    {
+        // using BigDecimal to handle large file system
+        BigDecimal used = BigDecimal.ZERO;
+        BigDecimal total = BigDecimal.ZERO;
+
+        for (Directories.DataDirectory dir : dataDirectoriesSupplier.get())
+        {
+            used = used.add(BigDecimal.valueOf(dir.getSpaceUsed()));
+            total = total.add(BigDecimal.valueOf(dir.getTotalSpace()));
+        }
+
+        used = used.add(BigDecimal.valueOf(getAllMemtableSize()));
+
+        if (logger.isTraceEnabled())
+            logger.trace("Disk Usage Guardrail: current disk usage = {}, total disk usage = {}.",
+                         FileUtils.stringifyFileSize(used.doubleValue()),
+                         FileUtils.stringifyFileSize(total.doubleValue()));
+
+        return used.divide(total, 5, BigDecimal.ROUND_UP).doubleValue();
+    }
+
+    @VisibleForTesting
+    public long getAllMemtableSize()
+    {
+        long size = 0;
+
+        for (String keyspaceName : Schema.instance.getKeyspaces())
+        {
+            Keyspace keyspace = Keyspace.open(keyspaceName);
+
+            for (ColumnFamilyStore cfs : keyspace.getColumnFamilyStores())
+            {
+                for (Memtable memtable : cfs.getTracker().getView().getAllMemtables())
+                {
+                    size += memtable.getLiveDataSize();
```

Comment:
Seems odd to include memtable size since the memtable isn't compressed.  It might be a good idea to multiply this by the average compression ratio of the table to get the actual on disk size.

Reply:
that's how it is in the original implementation. I'll try and follow up with the original author